mediacloud/mediawords/db/result/result.py

```diff
@@ -128,28 +143,44 @@ def array(self) -> List[Any]:
         row_tuple = self.__cursor.fetchone()
         if row_tuple is not None:
             row = list(row_tuple)
+
+            row = [self.__convert_datetime_objects_to_strings(_) for _ in row]
```

Comment:
should be able to get rid of the `row = list(row_tuple)` line.  Could you also use a different identifier from `_`?

Reply:
Got rid of `_` because we're not throwing the value away, thanks for the note.

`fetchone()` returns a psycopg2's `DictRow` object which is something between a `list` and a `dict`; here, we're being explicit about which one is the one that we want in each case. Also, Perl-Python binding code deep-copies `dict`s / `list`s returned by this code for Perl to be able to modify them afterwards; with `DictRow` being returned, it won't be able to do that deep copy because it's neither of those primitive types.

In our Perl-Python interactions, we [limit the data types being passed between the languages](https://github.com/berkmancenter/mediacloud/blob/master/doc/perl_to_python_rewrite.mdown#dont-call-perl-code-from-python) to primitive types and objects of classes that we own. Passing around exotic stuff like `DictRow` would introduce some weird behavior somewhere.

Reply:
`[j for j in iterable]` is equivalent to `list(iterable)`.  Right now you iterate over `row_tuple` twice. The same goes for the data structures in the other comments.

Reply:
When we remove that row (which we intend to do after the rewrite), no one will remember to re-add the `dict()` / `list()` cast so the method will start returning `DictRow`s (unless `typing` does its job properly and warns us about this).

Reply:
I am not sure how to have a productive discussion about writing bad code in the future, but just wanted to point out that this is not good code now.

Reply:
Extra code that supports the Perl -> Python move (usually annotated with `MC_REWRITE_TO_PYTHON` comment) is designed to be easily removable. For example, in Perl the distinction between strings and integers is way more lax, so in some cases we iterate over a list passed from Perl and cast its values to int:

```python
    # MC_REWRITE_TO_PYTHON: cast strings to ints
    # noinspection PyTypeChecker
    stories_ids = [int(x) for x in stories_ids]
```

Also, as discussed in #287 already, Perl wrapper around the database handler does a deep recursive copy of every object returned by the underlying psycopg2 handler because in the code, we modify some variables right after returning them, and results returned by the Perl-Python glue C code are immutable. The deep-copy part is also a one liner:

```perl
    sub hashes($)
    {
        my $self           = shift;
        my $return_value   = $self->{ _python_result }->hashes();
        my $writable_value = python_deep_copy( $return_value );
        return wantarray ? @{ $writable_value } : $writable_value;
    }
```

(No `MC_REWRITE_TO_PYTHON` annotation here because `python_deep_copy()` is easy to grep for.)

The whole idea behind keeping the business logic and the Perl-Python integration bits separated (even at the cost of the efficiency) is that after the rewrite, one could simply grep the whole codebase for `MC_REWRITE_TO_PYTHON` and remove the annotated blocks of code without much of the brainpower involved. The Python rewrite itself is hard enough, and there's a risk of having Perl memorabilia in the Python codebase forever after the rewrite, so we want the subsequent cleanup to be as easy as we can make it.

Now, in this case, if we have:

```python
# Business logic; iterates over list
whatever = list(whatever)

# Perl-Python rewrite shim; iterates over the list again
whatever = [foo(x) for x in list]
```

and we optimize it into a single iteration (to save oh-so-precious CPU cycles on a machine with 32 cores):

```python
# Business logic plus Perl-Python rewrite shim
whatever = [foo(x) for x in list]
```

we will no longer be able to just blindly remove the Perl-Python rewrite shim after the rewrite because the business logic (however small it is) is now merged into said shim, and one has to think about how to rewrite it - using just the *Backspace* key for cleaning up the Perl-Python glue is no longer enough.

Reply:
I'm not sure what you were trying to get across with this! Merge if you would like, just pointing out there is duplicated and inefficient code.

> usually annotated with MC_REWRITE_TO_PYTHON comment

None of the suggested changes are annotated with this

> to save oh-so-precious CPU cycles on a machine with 32 cores

I am not sure how that helps or changes the conversation?  You are still doing twice as much work as you need to.